tests/test_flows.py

```diff
@@ -0,0 +1,356 @@
+        self.repeater2_ucc.dev_send_1905(self.mac_gateway, 0x800C,
+            (0x94,0x0010,"{%s} 0x11 0x1002 0x90 0x1c2c3c 0x1d2d3d" % self.mac_repeater2_wlan2),
+            (0x96,0x001a,"{77:44:33:22:11:00} 0x01 {%s} 0x19293949 0x10203040 0x11213141 0x99" % self.mac_repeater2_wlan2),
+            (0xa2,0x0022,"{77:44:33:22:11:00} 0xa0203040 0xa1213141 0xa2223242 0xa3233343 "
+                         "0xa4243444 0xa5253545 0xa6263646"))
+        self.check_log(self.gateway, "controller", "Received AP_METRICS_RESPONSE_MESSAGE")
+
+        self.debug("Send 1905 Link metric query to agent 1 (neighbor agent 2)")
+        self.gateway_ucc.dev_send_1905(self.mac_repeater1, 0x0005,
+                                       (0x08,0x0008,"0x01 {%s} 0x02" % self.mac_repeater2))
+        self.check_log(self.repeater1, "agent_wlan0", "Received LINK_METRIC_QUERY_MESSAGE")
+        # TODO agent should send response autonomously
+        self.repeater1_ucc.dev_send_1905(self.mac_gateway, 0x6,
+            (0x09,0x0029,"{%s} {%s} {%s} {%s} 0x0100 0x01 0x00000000 0x0000e300 0x4230 0x0064 0x0300" %
+                (self.mac_repeater1, self.mac_repeater2, self.mac_repeater1_wlan0, self.mac_repeater2_wlan2)),
+            (0x0a,0x0023,"{%s} {%s} {%s} {%s} 0x0100 0x00000007 0x00020000 0x31" %
+                (self.mac_repeater1, self.mac_repeater2, self.mac_repeater1, self.mac_repeater2)))
+        self.check_log(self.gateway, "controller", "Received LINK_METRIC_RESPONSE_MESSAGE")
+
+        # Trigger combined infra metrics
+        self.debug("Send Combined infrastructure metrics message to agent 1")
+        self.gateway_ucc.dev_send_1905(self.mac_repeater1, 0x8013)
+        self.check_log(self.repeater1, "agent", "Received COMBINED_INFRASTRUCTURE_METRICS")
+        self.check_log(self.repeater1, "agent", "Received TLV_TRANSMITTER_LINK_METRIC")
+        self.check_log(self.repeater1, "agent", "Received TLV_RECEIVER_LINK_METRIC")
+
+    def test_client_steering_mandate(self):
+        self.debug("Send topology request to agent 1")
+        self.gateway_ucc.dev_send_1905(self.mac_repeater1, 0x0002)
+        time.sleep(1)
+        self.debug("Confirming topology query was received")
+        self.check_log(self.repeater1, "agent", "TOPOLOGY_QUERY_MESSAGE")
+
+        self.debug("Send topology request to agent 2")
+        self.gateway_ucc.dev_send_1905(self.mac_repeater2, 0x0002)
+        time.sleep(1)
+        self.debug("Confirming topology query was received")
+        self.check_log(self.repeater2, "agent", "TOPOLOGY_QUERY_MESSAGE")
+
+        self.debug("Send Client Steering Request message for Steering Mandate to CTT Agent1")
+        self.gateway_ucc.dev_send_1905(self.mac_repeater1, 0x8014,
+            (0x9B, 0x001b, "{%s 0xe0 0x0000 0x1388 0x01 {0x000000110022} 0x01 {%s 0x73 0x24}}" %
+                (self.mac_repeater1_wlan0, self.mac_repeater2_wlan0)))
+        time.sleep(1)
```

Comment:
Probably better to add `time.sleep(1)` to the `dev_send_1905` command, instead of having to write it manually each time. People are going to forget to do that time and time again.
Also, I believe in python you can do `time.sleep(0.1)`, i.e. you don't need to wait for an integer number of seconds.

Reply:
This sleep is taken over directly from the existing test_flows.sh script, that's why I kept it.

I have added a sleep(.1) in check_log. With python, it is needed a lot more often because there's no fork/exec overhead unlike the shell script. I chose to do it in check_log instead of dev_send_1905 because my idea is to convert check_log into a wait-with-timeout, which will remove the need for sleeping again.